PHP 8.5 migration guide

mumumu opened the pull request:
Based on [PHP 8.5 UPGRADE NOTES](https://github.com/php/php-src/blob/PHP-8.5/UPGRADING) and past migration guides.

- TODO
  * [x] New Features
  * [x] New Classes and Interfaces
  * [x] New Functions
  * [x] New Global Constants
  * [x] Backward Incompatible Changes
  * [x] Deprecated Features
  * [x] Windows Support
  * [x] Other Changes
  * [x] doc-base/manual.xml.in change
    - https://github.com/php/doc-base/pull/258

TimWolla commented:
https://github.com/php/php-src/pull/20278

Girgias commented on appendices/migration85/deprecated.xml:
```diff
@@ -0,0 +1,542 @@
+   <simpara>
+   <link linkend="language.operators.execution">The backtick operator</link>
+   as an alias for <function>shell_exec</function> has been deprecated.
+   </simpara>
```
Nit: indentation seems to be missing for the content of the simpara?

mumumu replied:
inentation fixed in https://github.com/php/doc-en/pull/4906/commits/8b62bd4cc5d44379f31b2b5d7033d1e08daed32a.

Girgias commented on appendices/migration85/deprecated.xml:
```diff
@@ -0,0 +1,542 @@
+   <!-- RFC: https://wiki.php.net/rfc/deprecations_php_8_5#deprecate_using_values_null_as_an_array_offset_and_when_calling_array_key_exists -->
+
+   <simpara>
+    Using null as an array offset or when calling <function>array_key_exists</function>
```
```suggestion
    Using &null; as an array offset or when calling <function>array_key_exists</function>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   <title>Destructing non-array values</title>
+
+   <simpara>
+    Destructing non-array values (other than NULL) using [] or list() now
```
```suggestion
    Destructing non-array values (other than &null;) using <literal>[]</literal> or <function>list</function> now
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+  <para>
+   <function>bzcompress</function> now throws a <classname>ValueError</classname>
+   when <parameter>$block_size</parameter> is not between 1 and 9.
+  </para>
```
```suggestion
  <simpara>
   <function>bzcompress</function> now throws a <classname>ValueError</classname>
   when <parameter>$block_size</parameter> is not between 1 and 9.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   contains nul bytes.
+   This aligns the type of Error thrown to be consistent with the rest of
+   the language.
+  </para>
```
```suggestion
  <simpara>
   <function>finfo_file</function> and <methodname>finfo::file</methodname>
   now throws a <exceptionname>ValueError</exceptionname> instead of a
   <exceptionname>TypeError</exceptionname> when <parameter>$filename</parameter>
   contains nul bytes.
   This aligns the type of Error thrown to be consistent with the rest of
   the language.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+
+  <para>
+   Calling the mysqli constructor on an already-constructed object
+   is now no longer possible and throws an Error.
```
```suggestion
   is now no longer possible and throws an <exceptionname>Error</exceptionname>.
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+  </para>
+
+  <para>
+   A <classname>ValueError</classname> is now thrown if
```
```suggestion
   A <exceptionname>ValueError</exceptionname> is now thrown if
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+
+   <simpara>
+    <methodname>IntlDateFormatter::setTimeZone</methodname>/<function>datefmt_set_timezone</function>
+    throws an <classname>IntlException</classname> on uninitialised
```
```suggestion
    throws an <exceptionname>IntlException</exceptionname> on uninitialised
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+   <simpara>
+    <function>session_start</function> is stricter in regard to the options
+    argument. It throws a <classname>ValueError</classname> if the array is
+    not a hashmap, or a <classname>TypeError</classname> if the read_and_close
```
```suggestion
    argument. It throws a <exceptionname>ValueError</exceptionname> if the array is
    not a hashmap, or a <exceptionname>TypeError</exceptionname> if the read_and_close
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+    <function>snmp3_get</function>,
+    <function>snmp3_set</function>
+    and <methodname>SNMP::__construct</methodname> throw a
+    <classname>ValueError</classname> when the hostname
```
```suggestion
    <exceptionname>ValueError</exceptionname> when the hostname
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+   <title>OpenSSL</title>
+
+   <simpara>
+    Added openssl.libctx to select the OpenSSL library context type. Either
```
Link for new INI setting TODO comment?

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+    <methodname>tidy::parseString</methodname> now throws a
+    <classname>ValueError</classname> if the configuration contains an
+    invalid value or attempts to set a read-only internal entry,
+    and a <classname>TypeError</classname> if a configuration key is not a
```
```suggestion
    <exceptionname>ValueError</exceptionname> if the configuration contains an
    invalid value or attempts to set a read-only internal entry,
    and a <exceptionname>TypeError</exceptionname> if a configuration key is not a
```

Girgias commented on appendices/migration85/new-features.xml:
```diff
@@ -0,0 +1,497 @@
+    The <link linkend="language.oop5.cloning">clone language construct</link>
+    is now a function and supports reassigning (readonly) properties during
+    cloning via the new <property>$withProperties</property> parameter.
+    <!-- RFC: https://wiki.php.net/rfc/clone_with_v2 -->
```
@TimWolla it cannot be disabled via disable_functions right?

Girgias commented on appendices/migration85/new-features.xml:
```diff
@@ -0,0 +1,497 @@
+   <methodname>XSLTProcessor::setParameter</methodname> and
+   <methodname>XSLTProcessor::removeParameter</methodname> now actually works
+   instead of being treated as empty.
+   This only works if the $name argument does not use Clark notation and is not a
```
```suggestion
   This only works if the <parameter>$name</parameter> argument does not use Clark notation and is not a
```

Girgias reviewed (commented):
Thank you for starting this! I left a lot of comments that I would like to have addressed. :)

mumumu commented on appendices/migration85/deprecated.xml:
```diff
@@ -0,0 +1,542 @@
+   The <constant>MHASH_<replaceable>*</replaceable></constant> constants
+   have been deprecated.  These have been overlooked when the mhash*() function
+   family has been deprecated per
+   <!-- RFC: https://wiki.php.net/rfc/deprecations_php_8_1#mhash_function_family -->
```
ok. deleted `mhash*()` related description in https://github.com/php/doc-en/pull/4906/commits/dc23d16030d94dc42e45990667c761e1ae827602.

mumumu commented on appendices/migration85/deprecated.xml:
```diff
@@ -0,0 +1,542 @@
+   and <classname>DatePeriod</classname> has been deprecated in favour of
+   the <link linkend="object.unserialize">__unserialize()</link> magic method.
+   <!-- RFC: https://wiki.php.net/rfc/deprecations_php_8_5#deprecate_the_sleep_and_wakeup_magic_methods -->
+  </simpara>
```
ok. deleted in https://github.com/php/doc-en/pull/4906/commits/eaa8c6828f773ee65b258fec0e560a94ce67f169.

mumumu commented:
@Girgias 

Thanks for all the feedback! I've addressed the points you raised, but I'm not sure how to fix the following.

https://github.com/php/doc-en/pull/4906/files#r2473529673

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+
+  <para>
+   Unicode data tables have been updated to Unicode 17.0
+  </para>
```
```suggestion
  <simpara>
   Unicode data tables have been updated to Unicode 17.0
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+  <para>
+   <function>ldap_get_option</function> and <function>ldap_set_option</function>
+   now throw a <exceptionname>ValueError</exceptionname> when passing an invalid option.
+  </para>
```
```suggestion
  <simpara>
   <function>ldap_get_option</function> and <function>ldap_set_option</function>
   now throw a <exceptionname>ValueError</exceptionname> when passing an invalid option.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   <function>pcntl_exec</function> now throws <exceptionname>ValueError</exceptionname>
+   when entries or keys of the <parameter>$env_vars</parameter> parameter
+   contain null bytes.
+  </para>
```
```suggestion
  <simpara>
   <function>pcntl_exec</function> now throws <exceptionname>ValueError</exceptionname>
   when entries of the <parameter>$args</parameter> parameter contain null bytes.
  </simpara>

  <simpara>
   <function>pcntl_exec</function> now throws <exceptionname>ValueError</exceptionname>
   when entries or keys of the <parameter>$env_vars</parameter> parameter
   contain null bytes.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   about this is now emitted.
+   To pass a variable by-ref to a constructor argument use the general
+   array value reference assignment: <code>$ctor_args = [&amp;$valByRef]</code>
+  </para>
```
```suggestion
  <simpara>
   The constructor arguments set in conjunction with
   <constant>PDO::FETCH_CLASS</constant> now follow the usual CUFA
   (<function>call_user_func_array</function>) semantics.
   This means string keys will act like a named argument.
   Moreover, automatic wrapping for by-value arguments passed to a by-ref
   parameter has been removed, and the usual <constant>E_WARNING</constant>
   about this is now emitted.
   To pass a variable by-ref to a constructor argument use the general
   array value reference assignment: <code>$ctor_args = [&amp;$valByRef]</code>
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   <constant>PDO::FETCH_INTO</constant> is used as a fetch mode in
+   <methodname>PDO::fetchAll</methodname>,
+   similar to <constant>PDO::FETCH_LAZY</constant>.
+  </para>
```
```suggestion
  <simpara>
   A <exceptionname>ValueError</exceptionname> is now thrown if
   <constant>PDO::FETCH_INTO</constant> is used as a fetch mode in
   <methodname>PDO::fetchAll</methodname>,
   similar to <constant>PDO::FETCH_LAZY</constant>.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   Passing an XPath expression that returns something other than a node set
+   to <methodname>SimpleXMLElement::xpath</methodname> will now emit a warning
+   and return &false;, instead of silently failing and returning an empty array.
+  </para>
```
```suggestion
  <simpara>
   Passing an XPath expression that returns something other than a node set
   to <methodname>SimpleXMLElement::xpath</methodname> will now emit a warning
   and return &false;, instead of silently failing and returning an empty array.
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   method, while the new backends will be used when passing either
+   <classname>Uri\Rfc3986\Uri</classname> or <classname>Uri\WhatWg\Url</classname>.
+   <!-- RFC: https://wiki.php.net/rfc/url_parsing_api#plugability -->
+  </para>
```
```suggestion
  <simpara>
   <methodname>SoapClient::__doRequest</methodname> now accepts a new,
   optional <parameter>$uriParserClass</parameter> parameter accepting
   string or &null; arguments.
   &null; represents the original (<function>parse_url</function>) based
   method, while the new backends will be used when passing either
   <classname>Uri\Rfc3986\Uri</classname> or <classname>Uri\WhatWg\Url</classname>.
   <!-- RFC: https://wiki.php.net/rfc/url_parsing_api#plugability -->
  </simpara>
```

Girgias commented on appendices/migration85/incompatible.xml:
```diff
@@ -0,0 +1,452 @@
+   Using a printf-family function with a formatter that did not specify the
+   precision previously incorrectly reset the precision instead of treating
+   it as a precision of 0.
+  </para>
```
```suggestion
  <simpara>
   Using a printf-family function with a formatter that did not specify the
   precision previously incorrectly reset the precision instead of treating
   it as a precision of 0.
  </simpara>
```

Girgias commented on appendices/migration85/new-features.xml:
```diff
@@ -0,0 +1,497 @@
+   <constant>PDO\Sqlite::TRANSACTION_MODE_IMMEDIATE</constant>,
+   and <constant>PDO\Sqlite::TRANSACTION_MODE_EXCLUSIVE</constant>,
+   allowing to configure the transaction mode to use when calling beginTransaction().
+  </simpara>
```
```suggestion
  <simpara>
   Added class constant <constant>Pdo\Sqlite::ATTR_BUSY_STATEMENT</constant>.
  </simpara>

  <simpara>
   Added class constants <constant>Pdo\Sqlite::ATTR_EXPLAIN_STATEMENT</constant>,
   <constant>Pdo\Sqlite::EXPLAIN_MODE_PREPARED</constant>,
   <constant>Pdo\Sqlite::EXPLAIN_MODE_EXPLAIN</constant>,
   <constant>Pdo\Sqlite::EXPLAIN_MODE_EXPLAIN_QUERY_PLAN</constant>.
  </simpara>

  <simpara>
   Added <constant>Pdo\Sqlite::ATTR_TRANSACTION_MODE</constant> connection attribute
   with possible values <constant>Pdo\Sqlite::TRANSACTION_MODE_DEFERRED</constant>,
   <constant>Pdo\Sqlite::TRANSACTION_MODE_IMMEDIATE</constant>,
   and <constant>Pdo\Sqlite::TRANSACTION_MODE_EXCLUSIVE</constant>,
   allowing to configure the transaction mode to use when calling beginTransaction().
  </simpara>
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+    <function>transliterator_get_error_message</function>,
+    <methodname>TransLiterator::getErrorCode</methodname>,
+    and <methodname>TransLiterator::getErrorMessage</methodname>
+    have dropped the false from the return type union. Returning &false;
```
```suggestion
    have dropped &false; from the return type union. Returning &false;
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+   <title>PDO_PGSQL</title>
+
+   <simpara>
+    <methodname>PDO::pgsqlCopyFromArray</methodname> now supports Iterable inputs.
```
```suggestion
    <methodname>Pdo\Pgsql::copyFromArray</methodname> now supports <type>iterable</type> inputs.
```

Girgias commented on appendices/migration85/other-changes.xml:
```diff
@@ -0,0 +1,751 @@
+   <simpara>
+    <methodname>PDO::sqliteCreateCollation</methodname> will now throw an
+    exception if the callback has the wrong return type, making it more
+    in line with <methodname>Pdo_Sqlite::createCollation</methodname> behavior.
```
```suggestion
    in line with <methodname>Pdo\Sqlite::createCollation</methodname> behavior.
```

Girgias reviewed (approved):
Sorry for the delay haven't been feeling well.

Lets just merge this now and do more improvements over time.
